### PR TITLE
feat(inputs): Add SearchInput and TextField

### PR DESCRIPTION
**Pull request by Anto2441**
## Description

- Refactor **`Input`** component to make it usable in **`SearchInput`** & **`TextField`**
- Add a new component **`SearchInput`** with his stories
- Add a new component **`TextField`** with his stories

Capture d’écran 2024-03-26 à 15 53 53

### CODE REVIEW

File: src/shared/ui/inputs/Input/Input.tsx [hunk 1]
```diff
@@ -1,46 +1,22 @@
...
+export const Input = React.forwardRef<HTMLInputElement, InputProps>(
+  function Input({ className, onChange, placeholder, type, ...props }, ref) {
+    return (
+      <input
+        {...props}
+        ref={ref}
+        className={className}
+        onChange={onChange}
+        placeholder={placeholder}
+        type={type}
```
Comment (carlo-ferreira) [inline src/shared/ui/inputs/Input/Input.tsx]: I think we can remove them. They are included in the `{...props}`. Is is useful to declare them when you want to add default behaviour for a props. For example for className:

```
twJoin(`bg-red-500`, className)
```

> Ok, by default the element has a red background, but I allow consumers to override it or add more classes if needed

File: src/shared/ui/inputs/Input/Input.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import type { Meta, StoryObj } from "@storybook/react";
+
+import { Input } from "./Input";
+
+const meta = {
+  args: {
+    className:
+      "block min-h-control-size-base w-full rounded-border-radius-base border border-essential-subdued bg-background-base py-spacing-tighter-2 font-semibold shadow-essential-base placeholder:text-color-subdued hover:border-essential-base focus:border-essential-base focus:border-white focus:ring-white",
```
Comment (carlo-ferreira) [inline src/shared/ui/inputs/Input/Input.stories.tsx]: If feels to be a lot of classes the consumer needs to pass each time he wants to use an input. Do you think we can provide some good defaults to the `<Input />`? Maybe all input have a border? or a `font-semibold`? or a `placeholder:text-color-subdued`?

(I don't know really ^^. We should check different usage in the Spotify app and extract some good defaults if there are some 👍)

File: src/shared/ui/inputs/SearchInput/SearchInput.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  base: "block w-full border-0 bg-background-tinted-base pl-spacing-looser-3 shadow-essential-base placeholder:text-font-size-smaller",
+
+  variants: {
+    rounded: {
+      full: "rounded-full border-border-width-hairline border-transparent py-spacing-tighter placeholder:text-essential-subdued hover:border-essential-subdued hover:bg-background-tinted-highlight focus:border-border-width-hairline focus:border-essential-base focus:border-white focus:ring-white",
+      square:
+        "rounded-border-radius-base py-spacing-tighter-5 placeholder:text-color-subdued focus:ring-transparent",
+    },
+  },
+});
```
Comment (carlo-ferreira) [inline src/shared/ui/inputs/SearchInput/SearchInput.tsx]: Great to have used cva ✌️ !

If I may, I would split the `rounded` variant with at least a `size` one:

- The `rounded` variant is responsible for the radius shape (`full` or `square` eg: `rounded-full` vs `rounded-border-radius-base`)
- The `size` variant would be responsible of… the size of the input (eg: `py-spacing-tighter` vs py-spacing-tighter-5).

This way, we could have large square button (or small) and large rounded button (or small).

What do you think?

File: src/shared/ui/inputs/SearchInput/SearchInput.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  id,
+  onChange,
+  placeholder,
+  rounded,
+}: SearchInputProps & VariantProps<typeof searchInputVariants>) {
+  return (
+    <div className="relative">
+      <div className="pointer-events-none absolute inset-y-0 left-spacing-tighter-2 flex items-center pr-spacing-tighter">
+        {icon}
+      </div>
```
Comment (carlo-ferreira) [inline src/shared/ui/inputs/SearchInput/SearchInput.tsx]: Any reason we chose to pass an `icon` prop? Won't it always be the same magnifying glass icon?

File: src/shared/ui/inputs/SearchInput/SearchInput.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        "rounded-border-radius-base py-spacing-tighter-5 placeholder:text-color-subdued focus:ring-transparent",
+    },
+  },
+});
+
+interface SearchInputProps {
+  icon: React.ReactNode;
+  id: string;
+  placeholder: string;
+  onChange?: React.ChangeEventHandler<HTMLInputElement>;
```
Comment (carlo-ferreira) [inline src/shared/ui/inputs/SearchInput/SearchInput.tsx]: 🚀  Comments here would be great

File: src/shared/ui/inputs/TextField/TextField.stories.tsx
Comment (carlo-ferreira) [inline src/shared/ui/inputs/TextField/TextField.stories.tsx:1]: Thank you 🙏 👼

File: src/shared/ui/inputs/SearchInput/SearchInput.stories.tsx
Comment (carlo-ferreira) [inline src/shared/ui/inputs/SearchInput/SearchInput.stories.tsx:1]: Storybook ✅

File: src/shared/ui/inputs/Input/Input.tsx (hunk 1 above)
Comment (Anto2441) [reply src/shared/ui/inputs/Input/Input.tsx]: Good point, I deleted them except classname, as you can see below.

File: src/shared/ui/inputs/Input/Input.stories.tsx (hunk 2 above)
Comment (Anto2441) [reply src/shared/ui/inputs/Input/Input.stories.tsx]: From what I can see on the site, there are two types of input, and the one with the focus on the white borders comes up most often.

I think it's ok that this is the default style for input, let me know what you think!

File: src/shared/ui/inputs/SearchInput/SearchInput.tsx (hunk 3 above)
Comment (Anto2441) [reply src/shared/ui/inputs/SearchInput/SearchInput.tsx]: Good point ! I added the `size` variant.

File: src/shared/ui/inputs/SearchInput/SearchInput.tsx (hunk 4 above)
Comment (Anto2441) [reply src/shared/ui/inputs/SearchInput/SearchInput.tsx]: You're right, a props icon is not necessary here.

The magnifyng glass is now directly in the component.

File: src/shared/ui/inputs/SearchInput/SearchInput.tsx (hunk 5 above)
Comment (Anto2441) [reply src/shared/ui/inputs/SearchInput/SearchInput.tsx]: Added !

**Review by carlo-ferreira — APPROVED** [review]
Looks good to me, thank you 🚀

**Pull request merged**